File: backend/src/main/java/br/ufpr/sept/so2/shared/domain/valueobject/Email.java
package br.ufpr.sept.so2.shared.domain.valueobject;

import java.util.Objects;
import java.util.regex.Pattern;

public final class Email {

    private static final Pattern EMAIL_REGEX =
            Pattern.compile("^[a-zA-Z0-9._%+\\-]+@[a-zA-Z0-9.\\-]+\\.[a-zA-Z]{2,}$");

    private final String value;

    private Email(String value) {
        this.value = value;
    }

    public static Email of(String raw) {
        String normalized = raw == null ? "" : raw.trim().toLowerCase();
        if (!EMAIL_REGEX.matcher(normalized).matches()) {
            throw new IllegalArgumentException("Formato de email inválido: " + raw);
        }
        return new Email(normalized);
    }

    public boolean isInstitutional() {
        return value.endsWith("@ufpr.br");
    }

    public String getValue() {
        return value;
    }

    @Override
    public String toString() {
        return value;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        if (!(other instanceof Email email)) {
            return false;
        }
        return value.equals(email.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/shared/domain/valueobject/Email.kt
package br.ufpr.sept.so2.shared.domain.valueobject

data class Email private constructor(val value: String) {

    fun isInstitutional(): Boolean = value.endsWith("@ufpr.br")

    override fun toString(): String = value

    companion object {
        const val PATTERN = "^[a-zA-Z0-9._%+\\-]+@[a-zA-Z0-9.\\-]+\\.[a-zA-Z]{2,}$"
        const val PATTERN_OR_BLANK = "^$|$PATTERN"
        private val EMAIL_REGEX = Regex(PATTERN)

        @JvmStatic
        fun of(raw: String?): Email {
            val normalized = raw?.trim()?.lowercase().orEmpty()
            require(EMAIL_REGEX.matches(normalized)) {
                "Formato de email inválido: $raw"
            }
            return Email(normalized)
        }
    }
}
